const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,279 @@
+2. Adjust the Parser to support `const` modifiers before trait bounds
+3. Add an `is_const` field to the HIR's `TraitRef`
+4. Adjust lowering to pass through the `is_const` field from AST to HIR
+5. Add a a check to `librustc_typeck/check/wfcheck.rs` ensuring that all generic bounds
+    in an `impl const` block have the `in_const` flag set and all methods' `constness` field is
+    `Const`.
+6. Feature gate instead of ban `Predicate::Trait` other than `Sized` in
+    `librustc_mir/transform/qualify_min_const_fn.rs`
+7. Remove the call in https://github.com/rust-lang/rust/blob/f8caa321c7c7214a6c5415e4b3694e65b4ff73a7/src/librustc_passes/ast_validation.rs#L306
+8. Adjust the reference and the book to reflect these changes.
+
+## Const type theory
+
+This RFC was written after weighing practical issues against each other and finding the sweet spot
+that supports most use cases, is sound and fairly intuitive to use. A different approach from a
+type theoretical perspective started out with a much purer scheme, but, when exposed to the
+constraints required, evolved to essentially the same scheme as this RFC. We thus feel confident
+that this RFC is the minimal viable scheme for having bounds on generic parameters of const
+functions. The discussion and evolution of the type theoretical scheme can be found
+[here](https://github.com/rust-rfcs/const-eval/pull/8#issuecomment-452396020) and is only 12 posts
+and a linked three page document long. It is left as an exercise to the reader to read the
+discussion themselves.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It is not a fully general design that supports every possible use case,
+but it covers the most common cases. See also the alternatives.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Effect system
+
+A fully powered effect system can allow us to do fine grained constness propagation
+(or no propagation where undesirable). This is way out of scope in the near future
+and this RFC is forward compatible to have its background impl be an effect system.
+
+## Fine grained `const` annotations
+
+One could annotate methods instead of impls, allowing just marking some method impls
+as const fn. This would require some sort of "const bounds" in generic functions that
+can be applied to specific methods. E.g. `where <T as Add>::add: const` or something of
+the sort. This design is more complex than the current one and we'd probably want the
+current one as sugar anyway
+
+## No explicit `const` bounds
+
+One could require no `const` on the bounds (e.g. `T: Trait`) and assume constness for all
+bounds. An opt-out via `T: ?const Trait` would then allow declaring bounds that cannot be
+used for calling methods. This design causes discrepancies with `const fn` pointers as
+arguments (where the constness would be needed, as normal function pointers already exist
+as the type of constants). Also it is not forward compatible to allowing `const` trait bounds
```

Comment:
So the open question from the last few comments on this RFC is "can we do the following?"

```rust
fn foo<T: const Trait>() -> u32 {
    const FOO: u32 = T::some_method();
    FOO
}
```

Which is totally out of scope of this RFC, because we can't even do

```rust
fn foo<T: Trait>() -> u32 {
    const FOO: u32 = T::SomeAssocConst;
    FOO
}
```

(see [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cf4d2d436dc64fc274cc0c7e013ee1a8))

```
error[E0401]: can't use type parameters from outer function
 --> src/lib.rs:6:22
  |
5 | fn foo<T: Trait>() -> u32 {
  |    --- - type variable from outer function
  |    |
  |    try adding a local type parameter in this method instead
6 |     const FOO: u32 = T::SomeAssocConst;
  |                      ^^^^^^^^^^^^^^^^^ use of type variable from outer function
```

This boils down to the same question as whether we want to be able to support

```rust
fn foo(f: const fn() -> u32) {
    const FOO: u32 = f();
}
```

Which goes into the territory of generic value arguments and should be solved there instead of here.

Reply:
> Why does this relate to generic value arguments

The thing is that generic arguments of functions do not influence any items inside that function, unlike generic arguments of `impl` blocks, which allow using their generic arguments on their associated items. Regular function arguments (or const function arguments) being used in inner items boils down to the same thing (semantically) while being something utterly different syntactically and intuitively.

> why does that make this incompatible with const trait bounds on non-const functions?

It doesn't make it incompatible, these bounds just don't make any sense in the way the language is set up so far. I think an example demonstrating the problem more closely is

```rust
fn foo<T: Trait>() -> [u8; 3] {
    [42; T::SomeAssocConst]
}
```

This code will only compile if given a `T` argument whose `SomeAssocConst` constant is `3`. Such behavior is neither desirable (very duck-typy, because the bounds on the function don't show that this behavior is expected), nor easy to implement (is `T::FOO + 1` the same thing as `1 + T::Foo`? Or `2 * T::FOO` the same thing as `T::FOO + T::FOO`?)